### PR TITLE
This fixes a couple things...

### PR DESCRIPTION
**Pull request by BarryNolte**
The big change:
<s>
* Made the d2 grammar injectable into markdown documents, this causes syntax highlighting to work correctly based on d2 grammar.  Otherwise, embedded d2 code is not highlighted.
* Markdown within d2 documents now relies on the built-in markdown grammar instead of the local copy of the grammar.
</s>
Smaller changes:
* Added some color to the custom compile task
* Stripped the ' - to - ' from the completion message since it's not helpful to the user
* Fixed bugs #50,#55 
* Partial fix for #54

### CODE REVIEW

File: syntaxes/d2.tmLanguage.json
Comment (gavin-ts) [inline syntaxes/d2.tmLanguage.json:90]: this is what I'm getting testing with the files from #53

### before 
Screen Shot 2023-03-01 at 2 56 09 PM

### after
Screen Shot 2023-03-01 at 2 55 39 PM

**Review by gavin-ts — CHANGES_REQUESTED** [review]

File: pages/previewPage.html
Comment (BarryNolte) [reply pages/previewPage.html:148]: Debugging leftovers...

File: themes/dark-color-theme.json
Comment (BarryNolte) [reply themes/dark-color-theme.json:1]: Do we?  We don't supply any themes, which is a whole different category of extension.  Setting of individual syntax colors can be done with a bit more finesse.

Comment (gavin-ts) [reply themes/dark-color-theme.json:1]: hmm we currently import these which we use in the Terrastruct app, I think they are here because it is a vscode theme, but perhaps we should move it out of this repo. We originally had themes here: https://github.com/terrastruct/d2-vscode/blob/5ede180a9221366b93df28d29546cc58b4c79e9a/package.json#L53-L64 . 

thoughts @terrastruct/engineering ?

File: syntaxes/d2.tmLanguage.json
Comment (BarryNolte) [reply syntaxes/d2.tmLanguage.json]: I backed out my changes and have decided to leave well enough alone.  I can't confirm, but there might be a nested syntax limit in vscode with injected grammars.

File: syntaxes/d2.tmLanguage.json
Comment (gavin-ts) [inline syntaxes/d2.tmLanguage.json:176]: are these changes intentional?

File: src/tasks.ts
Comment (gavin-ts) [inline src/tasks.ts:61]:
```suggestion
      error = error.replace(" - to -", "");
```

File: src/taskRunner.ts
Comment (gavin-ts) [inline src/taskRunner.ts:106]: Hmm this message only appears if the preview isn't already open.

can we also return the error from d2Tasks.compile and pass it to this callback for the exact message?

I can find the error message in the terminal here: but d2 needs to add line numbers to this error for it to show up in the editor.
Screen Shot 2023-03-02 at 5 16 10 PM

File: pages/previewPage.html
Comment (gavin-ts) [inline pages/previewPage.html:155]:
```suggestion
      window.addEventListener("wheel", (event) => {
        if (event.ctrlKey === true) {
          event.preventDefault();
          if (event.deltaY > 0) {
            zoomMinus();
          } else {
            zoomPlus();
          }
        }
      }, { passive: false });
```

to prevent scrolling while zooming

this will also need something like #721 to render with dimensions set so zooming works correctly here

**Review by gavin-ts — CHANGES_REQUESTED** [review]
quite a few different changes in this PR but: 

- sketch: approve change, but couldn't reproduce issue
- zoom: see comment
- terminal colors: approve
- error message: see comment
- syntax: should these still be here?
- themes: hold off for discussion

File: syntaxes/d2.tmLanguage.json
Comment (BarryNolte) [reply syntaxes/d2.tmLanguage.json:176]: Yes, @nhooyr suggested getting rid of the extra $.

File: src/taskRunner.ts
Comment (BarryNolte) [reply src/taskRunner.ts:106]: This was intentional.  We leave the last 'good' compile up, but on the first 'show', if there are errors, there is no previous 'good' render to show.

At least a line number is required, and possibly a column, I thought I had a fix for this, but there are some bugs in vscode dealing with multiple problemMatchers, so I couldn't just add one that grabs the filename and error.  At the moment, one size has to fit all.

File: pages/previewPage.html
Comment (BarryNolte) [reply pages/previewPage.html:155]: I just upgraded d2 to v0.2.2 that changed the svg code and that threw off all my janky zoom code.  I'm working on it.

File: src/taskRunner.ts
Comment (gavin-ts) [reply src/taskRunner.ts:106]: ok that makes sense, but we want to have some indication of a problem if the terminal is collapsed. hmm can we give it something like line 1 col 1 if we don't have a line number, that way the error can still be shown in the editor?

File: pages/previewPage.html
Comment (gavin-ts) [reply pages/previewPage.html:155]: > I just upgraded d2 to v0.2.2 that changed the svg code and that threw off all my janky zoom code. I'm working on it.

ok tested a locally compiled d2 version that sets dimensions for svg and the zooming works with that

File: src/taskRunner.ts
Comment (BarryNolte) [reply src/taskRunner.ts:106]: I'll put the whole error message in.

File: pages/previewPage.html
Comment (BarryNolte) [reply pages/previewPage.html:155]: OK, so any fix I do will break on my next upgrade?

Comment (gavin-ts) [reply pages/previewPage.html:155]: If you make a fix now it should still work, you would just probably not use the set dimensions flag in the d2 call.

The idea is adding a flag that you could provide to set the dimensions on the svg, which would work with this zooming, so an option is waiting for that and just adding the flag to the d2 command.

**Comment by nhooyr** [comment]
- The themes are supplied by this extension, you can use them after installing. We don't want to have a second extension just for the themes.
- The markdown syntax is modified and different from the built in markdown grammar to avoid conflicting with d2's syntax. See docs at https://github.com/terrastruct/d2-vscode#markdowntmlanguagejson

File: src/taskRunner.ts
Comment (BarryNolte) [reply src/taskRunner.ts]: OK, fixed this up to better show the errors.  We could probably fix this even better with an overlay that pops up on error, but that is a nice to have vs must have.

Found a bug in our problemMatcher too, and that led to another vscode bug (https://github.com/microsoft/vscode/issues/133013#issue-995134699).  I'm thinking problemMatchers are a problem...

**Comment by BarryNolte** [comment]
>

**Comment by BarryNolte** [comment]
> * The themes are supplied by this extension, you can use them after installing. We don't want to have a second extension just for the themes.

The themes are completely unused.  They weren't complete, had no d2 specific colors and caused flashing problems when used (at least on my machine :) ).  If we want to colorize d2 tokens, there are better ways to achieve this than creating a theme which is a global change to vscode instead of a localized change to d2 related files.

**Comment by nhooyr** [comment]
> The themes are completely unused. They weren't complete, had no d2 specific colors and caused flashing problems when used (at least on my machine :) ). If we want to colorize d2 tokens, there are better ways to achieve this than creating a theme which is a global change to vscode instead of a localized change to d2 related files.

What do you mean they're unused? If you don't want to use the terrastruct theme you don't have to?

What highlight group did you find missing that makes you believe they're incomplete?
The themes aren't d2 specific they're just the same theme we use on the D2 editor at terrastruct.com.
We don't have any intention of having a d2 specific theme.

It's almost certainly not the terrastruct themes specifically causing the flashing problems on your machine. They're merely static files consumed by VS Code. Try switching to other themes and see if you can reproduce.

**Comment by BarryNolte** [comment]
This extension no longer uses them or presents them as a theme choice to the user.  If the D2 App uses them, then the files should be moved to the repository for the app.

**Comment by nhooyr** [comment]
> This extension no longer uses them or presents them as a theme choice to the user. If the D2 App uses them, then the files should be moved to the repository for the app.

Screenshot 2023-03-03 at 11 02 25 AM

As far as I can tell it does.

**Comment by BarryNolte** [comment]
> As far as I can tell it does.

What version is shown under the quickpick dropdown?  When I go to the extension page, it says 8.1 and doesn't have a theme choice after install.

**Comment by alixander** [comment]
yeah i saw it too on 0.3 and not on 0.8.1 anymore. why was it removed though? i liked the themes. doesn't seem like it's mutually exclusive to offer functionality and themes too.

**Comment by nhooyr** [comment]
That's weird, someone accidentally must have removed it. Will fix.

**Comment by nhooyr** [comment]
It was you @BarryNolte!

See https://github.com/terrastruct/d2-vscode/commit/bc5f8cb2f7ccabec594ce17be590919d38af1452

I fixed in master.

**Comment by BarryNolte** [comment]
Yes, my first PR. I would think time would be better spent on making the d2 token names work better with all existing themes instead of trying to reinvent the wheel.

**Comment by nhooyr** [comment]
> Yes, my first PR. I would think time would be better spent on making the d2 token names work better with all existing themes instead of trying to reinvent the wheel.

I don't think you've understood my previous comments.

These themes are required no matter, we figure we might as well share them publically too. There is nothing special about them except for the fact that they're used on terrastruct.com's D2 editor. You do not have to use them nor are they enabled automatically on install. Nor does anyone spend time on them except for our designer once a year maybe.

**Comment by gavin-ts** [comment]
I think the idea is no one is going to use these themes when these are just for the Terrastruct app and with

> Nor does anyone spend time on them

it doesn't make sense to include them when we could move them to the other repo.

**Comment by nhooyr** [comment]
> I think the idea is no one is going to use these themes when these are just for the Terrastruct app and with
> it doesn't make sense to include them when we could move them to the other repo.

I don't think it's fair of you guys to assume that no one uses them. You guys don't use them and that's fine. I actually do like and use the dark theme. It's just a nice easter egg that doesn't harm anything.

**Comment by gavin-ts** [comment]
ok so we are publishing these as themes for people to use and we're not just using it as a library to import then? I think this is what is unclear

**Comment by BarryNolte** [comment]
OK, I'm going to put my Program Manager hat on for a second.
* Do you use the theme while editing anything other than d2 code??
* Has anyone reported a bug in the theme?
* Has anyone (outside the company) said, 'hey, nice theme'?
Program Manager hat off...

**Comment by nhooyr** [comment]
Feel free to add some docs in the README @gavin-ts

**Comment by alixander** [comment]
aside from the themes stuff, which we'll table for now, is this ready to go? @BarryNolte @gavin-ts

File: package.json
Comment (gavin-ts) [inline package.json:46]: What was the problem matcher issue you found?

This change appears to incorrectly parse the error position

### before
Screen Shot 2023-03-13 at 4 45 49 PM

### after

Screen Shot 2023-03-13 at 4 39 16 PM

File: src/docToPreviewGenerator.ts
Comment (gavin-ts) [inline src/docToPreviewGenerator.ts:96]: nice

Screen Shot 2023-03-13 at 4 47 51 PM

**Review by gavin-ts — CHANGES_REQUESTED** [review]
> aside from the themes stuff, which we'll table for now, is this ready to go? @BarryNolte @gavin-ts

just have 2 things left I think: 
- fix error pattern regex
- revert themes deletion

**Comment by BarryNolte** [comment]
Works for me. If there's anything to clean up, I'd prefer to do it in a smaller PR, this one is a mess.

File: package.json
Comment (BarryNolte) [reply package.json:46]: Reverted my change, me and regex do not get along at all.

**Review by gavin-ts — CHANGES_REQUESTED** [review]
Just need `git checkout master -- themes` to get the json files back and `yarn lint && yarn prettier -w .` which should also fix the package.json merge conflicts

**Comment by BarryNolte** [comment]
Cross your fingers, I think it's good to go.

File: .gitignore
Comment (gavin-ts) [inline .gitignore:9]:
```suggestion
```

File: package.json
Comment (gavin-ts) [inline package.json:281]: this is a duplicate entry of below, I don't think you needed to manually resolve merge conflicts if you ran the prettier lint command on this.

File: package.json
Comment (gavin-ts) [inline package.json:281]: this doesn't seem right likely a merge issue, I'd just rollback to before merging and try running prettier as mentioned here:

> Just need `git checkout master -- themes` to get the json files back and (from root dir of repo) `yarn lint && yarn prettier -w .` which should also fix the package.json merge conflicts

**Review by gavin-ts — CHANGES_REQUESTED** [review]
I think it is best to revert back to before merging and try the 2 commands mentioned in last review. I think you won't need to merge anything if you do that

File: package.json
Comment (gavin-ts) [inline package.json:283]: menus.commandPalette above

File: package.json
Comment (gavin-ts) [inline package.json:280]:
```suggestion
    ]
  },
```

File: package.json
Comment (gavin-ts) [inline package.json:327]:
```suggestion
```

File: package.json
Comment (gavin-ts) [inline package.json:329]:
```suggestion
```

**Review by gavin-ts — CHANGES_REQUESTED** [review]
does `yarn prettier -w .` not work for your dev env?

**Comment by BarryNolte** [comment]
Apparently not for JSON files. There were a couple duplicate sections, that might be part of the reason for massive change.

**Comment by gavin-ts** [comment]
> Apparently not for JSON files. There were a couple duplicate sections, that might be part of the reason for massive change.

hmm ok I would double check it is running from the root directory (not and not from ./src). if that doesn't work could you apply the patch I posted above?

File: package.json
Comment (BarryNolte) [reply package.json]: OK, I looovvee subtle install bugs...  yarn is now working the way it's supposed to on my machine.

File: package.json
Comment (gavin-ts) [inline package.json:317]:
```suggestion
```

**Review by gavin-ts — CHANGES_REQUESTED** [review]
glad yarn is working now, just need 1 more prettier run for package.json after most recent merge

File: package.json
Comment (gavin-ts) [inline package.json:317]:
```suggestion
```

we don't use this as a direct dependency

File: package.json
Comment (gavin-ts) [inline package.json:316]: sadly I think this will be a prettier error
```suggestion
    "markdown-it-container": "^3.0.0"
```

**Review by gavin-ts — APPROVED** [review]
🎉

**Pull request merged**
